### PR TITLE
Kafka 6383: complete shut down for streams threads that have not started

### PR DESCRIPTION
**Pull request by rodesai**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by rodesai** [comment]
@guozhangwang @mjsax

**Comment by guozhangwang** [comment]
retest this please

**Comment by rodesai** [comment]
retest this please

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 1]
```diff
@@ -1088,7 +1094,11 @@ private long computeLatency() {
      */
     public void shutdown() {
         log.info("Informed to shut down");
-        setState(State.PENDING_SHUTDOWN);
+        State oldState = setState(State.PENDING_SHUTDOWN);
+        if (oldState == State.CREATED) {
+            // The thread may not have been started. Take responsibility for shutting down
+            this.completeShutdown(true);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java]: nit: remove `this`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 2]
```diff
@@ -1088,7 +1094,11 @@ private long computeLatency() {
      */
     public void shutdown() {
         log.info("Informed to shut down");
-        setState(State.PENDING_SHUTDOWN);
+        State oldState = setState(State.PENDING_SHUTDOWN);
+        if (oldState == State.CREATED) {
+            // The thread may not have been started. Take responsibility for shutting down
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:1099]: nit: `the thread never started, thus need to take responsibility for shutting down`

**Review by mjsax — COMMENTED** [review]
I am wondering, if we can design a test that forces the race condition we missed originally?

**Comment by guozhangwang** [comment]
> I am wondering, if we can design a test that forces the race condition we missed originally?

Hmm, that might be a bit tricky for a unit test since the race condition did not depend on the `Time` object even, but only depend on triggering start the threads, which is quite out-of-control and hard to mock. @rodesai do you have any ideas? If not I think we can leave it.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 2 above)
Comment (rodesai) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:1099]: It may have started but not been scheduled yet.

**Comment by rodesai** [comment]
It would be hard to test that exact scenario without instrumenting StreamsThread just for that test case (e.g. by putting a sleep before transitioning state in run() or having it wait on a condition fired after the test shuts the streams thread down). But I think its worth having a test that calls run() after shutdown() and verifies that run() doesn't try to shut down again. I'll add that.

**Review by mjsax — APPROVED** [review]

**Comment by mjsax** [comment]
@rodesai Test failure seems related.

**Comment by mjsax** [comment]
Retest this please.

**Comment by guozhangwang** [comment]
LGTM. Merged to trunk. Thanks for the patience @rodesai !

**Pull request closed without merging**